iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -106,84 +102,178 @@ void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const cxx::optional<capro::IdString_t>& instance,
                            const cxx::optional<capro::IdString_t>& event) const noexcept
 {
-    ServiceDescriptionVector_t serviceInstanceResult;
-    // Find (K1, K2)
-    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
-    if (instance && service)
+    constexpr Wildcard WILDCARD;
+
+    // all search cases have complexity O(n)
+    // there are 8 cases and we dispatch with as little case-checking as possible
+    if (service)
     {
-        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
-        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleServices;
-        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleInstances;
+        if (instance)
+        {
+            if (event)
+            {
+                return find(*service, *instance, *event, searchResult);
+            }
+            return find(*service, *instance, WILDCARD, searchResult);
+        }
+        if (event)
+        {
+            return find(*service, WILDCARD, *event, searchResult);
+        }
+        return find(*service, WILDCARD, WILDCARD, searchResult);
+    }
 
-        auto rangeServiceMap = m_serviceMap.equal_range(service.value());
-        for (auto entry = rangeServiceMap.first; entry != rangeServiceMap.second; ++entry)
+    if (instance)
+    {
+        if (event)
         {
-            possibleServices.push_back(entry->second);
+            return find(WILDCARD, *instance, *event, searchResult);
         }
+        return find(WILDCARD, *instance, WILDCARD, searchResult);
+    }
+    if (event)
+    {
+        return find(WILDCARD, WILDCARD, *event, searchResult);
+    }
+
+    return findAll(searchResult);
+}
+
+const ServiceRegistry::ServiceDescriptionVector_t ServiceRegistry::getServices() const noexcept
+{
+    ServiceDescriptionVector_t allEntries;
+    findAll(allEntries);
```

Comment:
~~Why is `findAll()` needed? You could also invoke the copy c'tor of `cxx::vector`.~~

Forgot that you are storing `cxx::optional`'s... maybe `getAll()` is better, because we're not really searching anything here.

Reply:
Initially it was another `find` overload but I wanted to be more expressive. It corresponds to all wildcard search.